src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,482 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <unordered_map>
+
+#include "ir/branch-utils.h"
+#include "ir/subtypes.h"
+#include "ir/type-updating.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "wasm-traversal.h"
+#include "wasm-type.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+struct Unsubtyping
```

Comment:
Some overview of the algorithm might help here. I spent a little time but it wasn't obvious to me why casts differ from other operations, for example, and hence why we need that double while loop on the work queue.

My lack of understanding may be because I am stuck in the general direction I thought could work, but perhaps can't (and so your solution might work but is more complex). I was thinking after you have gathered all the data from `noteSubtype` you are basically done, except from calling `propagateToSubTypes` (`struct-utils.h`). That is, whenever we see that a `B` was written to a place that is defined as `A` then `B` and all subtypes of `B` must be writable there, so they must be subtypes of `A`. What am I missing?

Reply:
If we have `C <: B <: A`, and we see that `B` is written to a location typed `A`, but `C` is never written to a `B` or `A` location, then your proposed solution would not be able to optimize anything, but the current implementation would be able to optimize `C` to not have any supertypes.

Casts do not introduce subtyping requirements on their own. If we have a cast from `A` to `C` (and no other casts) in the setup above, then it would be unnecessarily conservative to require `C <: A`, since we know that cast cannot succeed because we observed that `C` does not inhabit `A`.

I'm doing my best to get the most precise possible analysis (without making it a data flow analysis or anything) so we never have to come back and improve this ever again :)

Reply:
I see, thanks. Eager propagation doesn't do that well, good point.

About your second paragraph, I wonder if we can actually have a separate pass that finds situations like a cast from `A` to `C` that will never succeed because no `C` is written to an `A`. That pass could replace such casts with failures.

Assuming such a pass, would that not handle casts entirely? And then that would leave a simple operation without propagation: for each type T, we see where it is written, and make it a subtype of the closest ancestor to itself among those, and as a second stage add more subtyping to keep validation working.

Reply:
I think that could mostly work if we did a data flow analysis to find a set of types that can appear at each cast source, but even then, the data flow analysis would depend on subtyping information to be safe. For example, if a function parameter is a cast source, we would have to conservatively assume any of its subtypes could be the cast target unless we wanted to do an interprocedural analysis. No matter how precise we make the analysis, it still has to conservatively bottom out at depending on subtyping information.

So we have a circular dependency: we need subtyping information to analyze cast sources, and we need to analyze cast sources to get subtyping information. The only way to resolve that dependency optimally is to do both analyses in lockstep until reaching a fixed point, which is what the current implementation does.

Reply:
Thanks for explaining, I think I see the general shape of your algorithm now.

I still don't get what constraints casts place on us, however (aside from us needing to remove failing casts):

> For example, if a function parameter is a cast source, we would have to conservatively assume any of its subtypes could be the cast target unless we wanted to do an interprocedural analysis.

I don't see why we need that yet. Consider this module:

```wat
(module
  (type $A)
  (type $B (sub $A))

  (func $get-A (param $A)
    (ref.cast $B (local.get $A))
  )
)
```

Now let's add one of the following functions to it, starting with this:
```wat
(func $caller
)
```
We added nothing here. We can turn that cast into a trap, since we've never seen
any $B written to an $A location, and we can remove the subtyping between $B and
$A.

Next, imagine we added this function instead:
```wat
(func $caller
  (call $get-A (new $A))
)
```
The cast will fail at runtime. Again, we can see it will trap at compile time,
and remove the subtyping.

Finally, imagine we added this function instead:
```wat
(func $caller
  (call $get-A (new $B))
)
```
The cast will succeed at runtime. Our analysis will see that we write a $B to an
$A location, so $B must be a subtype of $A, so we can optimize nothing here.

Reply:
Hmm, I don't understand the testcase. I see a `local.set` of a `$bot.A` to `$top.A`, so we need that subtyping, and aside from that I see only casts, which as I said before, I don't think should actually constrain us (but we do need to fix up casts later). This is what I'd expect the optimization to emit for that module:
```wat
(module
 (rec
  (type $topC (sub (struct)))
  (type $midC (sub (struct)))
  (type $botC (sub (struct)))

  (type $topB (sub (struct (ref null $topC))))
  (type $midB (sub (struct (ref null $botC))))
  (type $botB (sub (struct (ref null $botC))))

  (type $topA (sub (struct (ref null $topB))))
  (type $midA (sub (struct (ref null $botB))))
  (type $botA (sub $topA (struct (ref null $botB)))) ;; only subtyping kept
 )

 (func $cast
  (local $l (ref null $topA))
  (local.set $l
   (struct.new $botA)
  )
  (drop
   (ref.cast (ref $botA) ;; rewritten to the only possibility, $botA
    (struct.new $topA)
   )
  )
  ;; Other two are outright impossible, so rewritten to trap.
  (unreachable)
  (unreachable)
 )
)
```
Is that output wrong somehow?

Reply:
That transformation happens to be correct for that test because the casts will fail, but we can construct cases where the semantics would change if we disregarded subtyping constraints arising from casts:
```
(module
 (type $top (sub (struct)))
 (type $mid (sub $top (struct)))
 (type $bot (sub $mid (struct)))
 
 (func $test
  (local $top (ref $top))
  (local $mid (ref $mid))
  (local $bot (ref $bot))
  (local.set $top (struct.new $bot)) ;; Requires $bot <: $top
  (local.set $mid (ref.cast (ref $mid) (local.get $top))) ;; Cast $top -> $mid
 )
)
```
The cast here should succeed because the value of type `(ref $bot)` inhabits `(ref $mid)`, but if you disregard the subtyping constraints that casts introduce, you end up with `$bot <: $top` as the only constraint, so you would rewrite the types like this:

```
(type $top (sub (struct)))
(type $mid (sub (struct))) ;; no longer related to $top or $bot
(type $bot (sub $top (struct)))
```

With that type rewriting, the module would still validate, but the cast would fail because the value of type `(ref $bot)` no longer inhabits `(ref $mid)`.

Here's the key correctness property for casts, as documented in `analyzeCasts`:
```
    // For each cast (src, dest) pair, any type that remains a subtype of src
    // (meaning its values can inhabit locations typed src) and that was
    // originally a subtype of dest (meaning its values would have passed the
    // cast) should remain a subtype of dest so that its values continue to pass
    // the cast.
```

Reply:
IIANM the last comment's output is maximally optimized, but over lunch I realized something that may be related to what you were trying to get me to understand before: while on that testcase what I'm saying can work , in other cases the "fix up casts" stuff that I've been handwaving can actually be difficult.

Specifically, say we have a cast to `$X` and there are two *sibling* types `$Y1, $Y2`. Then the cast could succeed if the input is either of those. I can think of three general approaches to that:

1. What I am guessing this PR does, which is to "reattach" `$Y1, $Y2` to `$X` as children (is that right?). Then you don't need to change the cast at all, but you do pay the cost of all of `$X`'s fields and their memory overhead.
2. Expand into casts: `(if (ref.test $Y1) (ref.cast $Y1) (ref.cast $Y2))`. This is larger in code size and may be slower, but it allows us to have any parent type for `$Y1, $Y2` so it could save memory. It is also "clean" in that we actually do remove the maximal amount of subtyping based on what is written where, and we can do that in a simple one-pass method. And then after that in another single pass we fix up all casts.
3. Slightly vague idea, but we can do a variation on the 1 and create `$shadow-X`, a "shadow hierarchy" of types that contain no fields and are only used for casts; `$Y1, $Y2` would be subtypes of the otherwise-empty `$shadow-X`, which has no memory overhead for them, and then we just need to make the cast be `(ref.cast $shadow-X)` rather than on `$X`. However, while this works for two siblings whose parents we want to remove, it's less obvious to me what to do if one is a subtype of `$X` anyhow and the other is not.

Reply:
(sorry, our comments raced; reading your comment now after posting mine)

Reply:
About the example from your last comment:
```wat
(module
 (type $top (sub (struct)))
 (type $mid (sub $top (struct)))
 (type $bot (sub $mid (struct)))
 
 (func $test
  (local $top (ref $top))
  (local $mid (ref $mid))
  (local $bot (ref $bot))
  (local.set $top (struct.new $bot)) ;; Requires $bot <: $top
  (local.set $mid (ref.cast (ref $mid) (local.get $top))) ;; Cast $top -> $mid
 )
)
```
In my idea (2 in my list) the output would be this:
```wat
(module
 (type $top (sub (struct)))
 (type $mid (sub (struct)))      ;; no super
 (type $bot (sub $top (struct))) ;; super improved
 
 (func $test
  (local $top (ref $top))
  (local $mid (ref $mid))
  (local $bot (ref $bot))
  (local.set $top (struct.new $bot))
  (local.set $mid (ref.cast (ref $bot) (local.get $top))) ;; cast type changed
 )
)
```
That almost works aside from the last `local.set`, so, yeah, as you said there is a problem here.

Separating the two problems, however, as I've suggested, we could first improve the cast, then optimize, and then remove supertypes. The optimization in the middle would get rid of `$mid`. That is, I think reusing the "what is written where" info from this pass to refine casts could help make this pass more effective. I was hoping it could also remove the cast handling complexity but that part might not be practical.

Reply:
I realize that `--abstract-type-refining` does what I was thinking about regarding casts, so we do already have that first part already :smile: 

Anyhow, I'm mostly convinced I should review this PR for landing, but let me know if anything in the last ideas suggests anything worthwhile.

Reply:
Yeah, optimizing casts and before and after this pass can be helpful, but we can't do it in the middle of this pass with incomplete type information.

Reply:
We do have enough to do idea 2 from earlier ("expand casts"), don't we? I agree there are tradeoffs there but we have all the information we need at least.

Reply:
Thinking more on this after we discussed it offline, I think I realized what was making it hard for me to understand before. I was expecting there to be a symmetry between all assignments, but there is actually a very unsymmetrical issue around casts. Say we have this:
```wat
;; $A :> B :> C
(local.set $A (new $B))
(local.set $B (new $C))
```
Then we have two constraints: B must be a subtype of A, and C of B. All we need is for data to be passable according to those rules, and in fact because of how subtyping works - it is recursive -  they imply `A :> B :> C`, so in particular C must be a subtype of A.

However, here:

```wat
;; $A :> B :> C
(local.set $A (new $C))
(cast $B (local.get $A))
```

Now we have two constraints, that C must be a subtype of A, and B must be a subtype of A. In theory we could fulfill those with both B and C being subtypes of A as siblings, but we need to be able to pass data around, and passing C to A and then that C-in-A to B will fail, because we no longer have that recursive property from before. That is, to be able to pass data around according to our two constraints, we need more than just those simple subtyping relations, but also some form of closure operation. If all we have are upcasts then we get it for free as we saw before; if we have downcasts then the symmetry is broken and we need more. Specifically we need everything that can reach A to be able to reach B, which implies subtyping between all subtypes of A to B.

A little verbose but maybe something like this in a comment in the pass could be helpful?